content/articles/spring-boot-recaptcha-v3/index.md

```diff
@@ -0,0 +1,855 @@
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.ToString;
+
+import javax.persistence.*;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@ToString
+@Entity(name = "role")
+@Table
+public class Role {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private Long id;
+
+    private String roleName;
+
+    public Role(String roleName) {
+        this.roleName = roleName;
+    }
+}
+
+```
+
+`@Data` - Generates all getters and setters for the fields in the class.
+
+`@AllArgsConstructor` - Generates a constructor with all the fields declared in the class.
+
+`@NoArgsConstructor` - Generates a constructor without any arguments.
+
+`@ToString` - Generates a string representation of the fields that can be used when debugging.
+
+
+Create a class named AppUser with fields `id`, `username`, and `password` which represents different users who can interact with the application depending on the rights they have.
+
+Since a user can have many rights we must include a Role field in the app user class and annotate with `@ManyToMany` annotation.
+
+```java
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.ToString;
+
+import javax.persistence.*;
+import java.util.ArrayList;
+import java.util.Collection;
+
+@Entity(name = "user")
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@ToString
+public class AppUser {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private Long id;
+
+    @Column(name = "user_name")
+    private String userName;
+
+    @Column(name = "password")
+    private String password;
+
+    @ManyToMany(fetch = FetchType.EAGER, cascade = CascadeType.ALL)
+    private Collection<Role> roles = new ArrayList<>();
+
+    public AppUser(String userName, String password) {
+        this.userName = userName;
+        this.password = password;
+    }
+}
+```
+Create a class named ReCaptchaResponse with field `success`, `hostname`, `action`, `score`, `challenge_ts`, and `errorCodes` which is an array of string and annotate it with `@JsonProperty("error-codes")` indicating that response from the server is a `JSON` and will be mapped to an array of strings. 
+
+```java
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Data;
+
+@Data
+public class ReCaptchaResponse {
+    private boolean success;
+    private String hostname;
+    private String action;
+    private float score;
+    private String challenge_ts;
+
+    @JsonProperty("error-codes")
+    private String[] errorCodes;
+}
+```
+
+### Create User and Role repositories
+Create an interface named `AppUserRepository` and extend `JpaRepository` which will provide us the capability to perform CRUD operations without writing any queries.
+
+Create a method that returns an `AppUser` by searching for a username and this can be achieved by adding the following method which is created by chaining the field names using special keywords.
+
```

Comment:
Change to "Create a method that returns an `AppUser` by searching for a username, and we can achieve this by adding the following method, created by chaining the field names using special keywords."